### DISCUSSION TITLE
Experiment to enable cardboard mode on iOS

### DISCUSSION THREAD

**Discussion by robinzhx** [mozilla/hubs#3210] [General Questions & Discussion] [answered]
Hi Everyone!

Is there any way to have a customized app allowing people with iOS be able to use a cheap cardboard solution to view Hubs in VR? We currently have a class of students that might need a quick solution. 

We host a Hubs Cloud in AWS, and I am willing to twist any some code in Hubs Cloud or develop a safari-like or WebXR iOS app for a small group of people to just test it using TestFlight. Is it even possible?

Best,
Robin

**Comment by robinkwilson** [comment, marked as answer]
@emclaren Thank you for the help with the answer! 

I'm afraid you may be roadblocked on the iOS/safari side -- and thus this might not be possible. You'd first enable the functionality on the phone. Then, on a custom client for Hubs Cloud, re-enable the view in cardboard option.

See this issue for context on A-Frame + Safari + Sensor access on phone: https://github.com/aframevr/aframe/issues/3976#issuecomment-458569087

**Reply by robinzhx** [reply]
Thanks @robin-k-wilson for the info! As I checked further, it seems to be able to re-enable by `DeviceMotionEvent.requestPermission` as how A-Frame now can access the sensor again. 
See this issue [Make dialog to call DeviceMotionEvent.requestPermission in iOS 13+](https://github.com/aframevr/aframe/issues/4287)
Is it ever gonna happened in the plan of further dev? Any option if I may help on developing this issue?

**Reply by brianpeiris** [reply]
I believe Hubs relies on a new enough version of a-frame that this may happen automatically. As I mentioned in my other comment, you could try enabling Cardboard support on Safari iOS in your custom Hubs Cloud client.
As for whether we'll be implementing this ourselves, I think nausea is still a concern for us, and additionally, support for Cardboard is usually pretty far down on our list of priorities.

**Comment by brianpeiris** [comment]
@robinzhx 

There are several considerations here. Firstly, we decided not to support VR mode on iOS because, at the time at least, VR mode in iOS Safari was essentially an emulated mode, where Hubs would have to use older gyroscope Web APIs to support VR mode.

The problem with these old gyroscope APIs is that they were never meant to be used for the type of realtime usage that VR requires. The gyroscope API is slow and not precise enough, and that cases a nauseating experience in a cardboard headset. As such, we decided to avoid putting users in that position.
You can try this for yourself on iOS with this aframe example: https://aframe.io/aframe/examples/test/gltf-model/

On Android, Google Chrome now supports WebXR APIs natively, and it uses native Google Cardboard APIs to enable that, which means the experience is much better.

In addition to native platform support, we generally haven't spent a lot of time improving Cardboard support, even on Android. At the moment, the only thing you can do in Cardboard mode is teleport around. None of the other features are available to cardboard users.

If you've got a custom hubs client deployed on your Hubs Cloud, you might want to try experimenting with enabling cardboard support on iOS Safari. You can start here, though I'm not sure if this is the only change you'll have to make: https://github.com/mozilla/hubs/blob/hubs-cloud/src/utils/vr-caps-detect.js#L65

To your question about a WebXR iOS app. This is obviously a much larger ask, so I'm not sure we can provide much support other than helping you understand the Hubs code. Anything related to the iOS app you'd have to build would be out of scope for us, as much as I would personally like to see that tried. 

One thing I'll note is that Mozilla did experiment with a WebXR browser on iOS, in case you were not aware. The main issue we ran into, with respect to Hubs, was that Apple does not allow these embedded browsers to access the Webkit microphone APIs and I believe WebRTC networking APIs may also be generally unavailable. If I'm not mistaken, it would mean you would have to build all of those capabilities from scratch as well. Take that with a grain of salt though, I'm not an expert in iOS development by any means.

Here are some links about Mozilla's WebXR viewer experiment. Unfortunately, as far as I know, this is now an unmaintained project. 
- Last release of the viewer: https://blog.mozvr.com/webxr-viewer-2-0-released/
- A repo for an older version of the WebXR viewer: https://github.com/mozilla-mobile/webxr-ios
- The last release had been implemented on top of the official Firefox for iOS code, but in a separate branch: https://github.com/mozilla-mobile/firefox-ios/tree/webxr

**Reply by robinzhx** [reply]
Really appreciate for such detailed explaination! I will try to enable through custom client through Hubs Cloud and see if that may work in certain way. About long run, it is defintely harder to implement low level APIs for iOS, I have tried the Mozilla WebXR viewer in App Store but unfortunately it doesn't support Hubs as well. I agreed Cardboard support should be lower priorities since Google abandon it soon and also new standard coming from Apple may soon hit the market this upcoming year. 

Much thanks!
